Wrappers/Python/ccpi/viewer/utils/io.py

```diff
@@ -42,3 +58,772 @@ def SaveRenderToPNG(render_window, filename):
+        reader.SetIsAcquisitionData(not self._ResampleZ)
+
+        if not self._Crop:
+            if self._Resample:
+                target_size = self._TargetSize
+            else:
+                # forced use of resample reader in the case that we
+                # don't want to crop or resample,
+                # but the large target size means we don't resample
+                target_size = 1e12
+            reader.SetTargetSize(int(target_size))
+
+        # Add observers:
+        reader.AddObserver(vtk.vtkCommand.ProgressEvent,
+                           partial(self._ReportProgress, progress_callback=progress_callback))
+
+        # Prints the error if an error occurs in the reader.
+        # Otherwise this wouldn't print at all.
+        error_obs = ErrorObserver(callback_fn=print)
+        reader.AddObserver(vtk.vtkCommand.ErrorEvent, error_obs)
+        # Could add end observer so that we don't continue to do anything
+        # else if an error does occur in reader?
+
+        return reader
+
+    def _GetMetaImageReader(self, progress_callback=None):
+        if self._Crop:
+            reader = cilMetaImageCroppedReader()
+            reader.SetTargetZExtent(tuple(self._TargetZExtent))
+        else:
+            reader = cilMetaImageResampleReader()
+        return reader
+
+    def _GetNumpyImageReader(self, progress_callback=None):
+        if self._Crop:
+            reader = cilNumpyCroppedReader()
+            reader.SetTargetZExtent(tuple(self._TargetZExtent))
+        else:
+            reader = cilNumpyResampleReader()
+
+        return reader
+
+    def _GetTiffImageReader(self, progress_callback=None):
+        if self._Crop:
+            reader = cilTIFFCroppedReader()
+            reader.SetTargetZExtent(tuple(self._TargetZExtent))
+        else:
+            reader = cilTIFFResampleReader()
+        return reader
+
+    def _GetRawImageReader(self):
+        if self._OriginalImageAttrs is None or 'shape' not in self._OriginalImageAttrs.keys():
+            raise Exception("To read a raw image, raw_image_attrs must be set.")
+
+        isFortran = self._OriginalImageAttrs['is_fortran']
+        isBigEndian = self._OriginalImageAttrs['is_big_endian']
+        typecode = self._OriginalImageAttrs['typecode']
+        shape = tuple(self._OriginalImageAttrs['shape'])
+
+        if self._Crop:
+            reader = cilRawCroppedReader()
+            reader.SetTargetZExtent(tuple(self._TargetZExtent))
+        else:
+            reader = cilRawResampleReader()
+
+        reader.SetBigEndian(isBigEndian)
+        reader.SetIsFortran(isFortran)
+        reader.SetTypeCodeName(typecode)
+        reader.SetStoredArrayShape(shape)
+
+        return reader
+
+    def _GetHDF5ImageReader(self):
+        if self._Crop:
+            reader = cilHDF5CroppedReader()
+            reader.SetTargetExtent([0, -1, 0, -1, self._TargetZExtent[0], self._TargetZExtent[1]])
```

Comment:
The HDF5 reader has differing behaviour with regards to setting the target extent - SetTargetExtent is used instead of SetTargetZExtent